### PR TITLE
Add Dell Enterprise SONiC 'pim_global' module

### PR DESCRIPTION
**Pull request by ArunSaravananBalachandran**
##### SUMMARY

- Add Enterprise SONiC 'pim_global' resource module:
  - Provide data definitions in 'argspec'.
  - Implement 'facts' to gather facts for the module.
  - Implement 'config' to apply configurations for merged, deleted, replaced and overridden states.
  - Implement support for 'check' and 'diff' modes.
- Define regression test cases to verify correct functionality of pim_global resource module.

##### GitHub Issues
List the GitHub issues impacted by this PR. If no Github issues are affected, please indicate this with "N/A".

| GitHub Issue # |
| -------------- |
| N/A |

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

- sonic_pim_global

##### Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have maintained backward compatibility or have provided any relevant "breaking_changes" descriptions in a "fragment" file in the "changelogs/fragments" directory of this repository.
- [ ] I have provided a summary for this PR in valid "fragment" file format in the "changelogs/fragments" directory of this repository branch. Reference : [Ansible Change Log Document](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to)

##### How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

Regression report: [pim_global_regression_report.pdf](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/15218783/pim_global_regression_report.pdf)

Regression report (check and diff modes): [pim_global_regression_check_diff_mode_report.pdf](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/15218788/pim_global_regression_check_diff_mode_report.pdf)

Regression terminal output (check and diff modes): [pim_global_regression_check_diff_mode_terminal_output.txt](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/15218795/pim_global_regression_check_diff_mode_terminal_output.txt)

### CODE REVIEW

**Review by kerry-meyer — CHANGES_REQUESTED** [review]
Thank you for providing this change set to implement the 'pim_global' resource module.

The code structure and test results look good.

I have posted a few comments regarding documentation in the 'module' file and some other comments/questions on implementation details.

There are also some merge conflicts introduced into the common/shared files by recent merges.

File: plugins/modules/sonic_pim_global.py [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+# ip pim vrf VrfReg1 join-prune-interval 120
+# ip pim vrf VrfReg1 keep-alive-timer 360
+# ip pim vrf default ecmp
+# sonic#
+
+
+"""
+RETURN = """
+before:
+  description: The configuration prior to the model invocation.
```
[the remark this answers is not in the document]
Comment (ArunSaravananBalachandran) [reply plugins/modules/sonic_pim_global.py]: Done.

File: plugins/modules/sonic_pim_global.py [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+  returned: always
+  type: list
+  sample: >
+    The configuration returned will always be in the same format
+     of the parameters above.
```
[the remark this answers is not in the document]
Comment (ArunSaravananBalachandran) [reply plugins/modules/sonic_pim_global.py]: Done.

File: plugins/modules/sonic_pim_global.py [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+after:
+  description: The resulting configuration model invocation.
+  returned: when changed
+  type: list
+  sample: >
+    The configuration returned will always be in the same format
+     of the parameters above.
```
[the remark this answers is not in the document]
Comment (ArunSaravananBalachandran) [reply plugins/modules/sonic_pim_global.py]: Done.

File: plugins/modules/sonic_pim_global.py [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+after(generated):
+  description: The generated configuration model invocation.
+  returned: when C(check_mode)
+  type: list
+  sample: >
+    The configuration returned will always be in the same format
+     of the parameters above.
```
[the remark this answers is not in the document]
Comment (ArunSaravananBalachandran) [reply plugins/modules/sonic_pim_global.py]: Done.

File: plugins/module_utils/network/sonic/config/pim_global/pim_global.py [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+    __DELETE_SUBCONFIG_AND_LEAFS,
+    get_formatted_config_diff,
+    get_new_config
+)
+from ansible.module_utils.connection import ConnectionError
+
+PATCH = 'patch'
+DELETE = 'delete'
+TEST_KEYS = [{'config': {'vrf_name': ''}}]
+TEST_KEYS_GENERATE_CONFIG = [{'config': {'vrf_name': '', '__delete_op': __DELETE_SUBCONFIG_AND_LEAFS}}]
```
[the remark this answers is not in the document]
Comment (ArunSaravananBalachandran) [reply plugins/module_utils/network/sonic/config/pim_global/pim_global.py]: As suggested, have made changes to use default (__DELETE_OP_DEFAULT).

File: plugins/module_utils/network/sonic/config/pim_global/pim_global.py [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+            url = self.pim_global_path.format(vrf=vrf)
+            vrf_requests = []
+            config_dict = {}
+            ssm_config_dict = {}
+
+            if 'ecmp_enable' in command:
+                if command['ecmp_enable']:
+                    config_dict['ecmp-enabled'] = command['ecmp_enable']
+                else:
+                    vrf_requests.append({'path': self.pim_global_config_path['ecmp_enable'].format(vrf=vrf), 'method': DELETE})
```
[the remark this answers is not in the document]
Comment (ArunSaravananBalachandran) [reply plugins/module_utils/network/sonic/config/pim_global/pim_global.py:288]: The delete is being done to be in line with performing the same configurations over CLI.

File: plugins/module_utils/network/sonic/config/pim_global/pim_global.py [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+
+            # Handle ECMP Rebalance configuration seperately
+            # 1) For enabling, configure it after enabling ECMP
+            # 2) For disabling, configure it before disabling ECMP
+            if 'ecmp_rebalance_enable' in command:
+                if command['ecmp_rebalance_enable']:
+                    payload = {'openconfig-network-instance:global': {'config': {'ecmp-rebalance-enabled': command['ecmp_rebalance_enable']}}}
+                    vrf_requests.append({'path': url, 'method': PATCH, 'data': payload})
+                else:
+                    vrf_requests.insert(0, {'path': self.pim_global_config_path['ecmp_rebalance_enable'].format(vrf=vrf), 'method': DELETE})
```
[the remark this answers is not in the document]
Comment (ArunSaravananBalachandran) [reply plugins/module_utils/network/sonic/config/pim_global/pim_global.py:313]: The delete is being done to be in line with performing the same configurations over CLI.

File: plugins/module_utils/network/sonic/facts/pim_global/pim_global.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+        :param spec: The facts tree, generated from the argspec
+        :param conf: The configuration
+        :rtype: dictionary
+        :returns: The generated config
+        """
+        config = deepcopy(spec)
+        config['vrf_name'] = vrf_name
+
+        if conf.get('config'):
+            config['ecmp_enable'] = conf['config'].get('ecmp-enabled', False)
```
[the remark this answers is not in the document]
Comment (ArunSaravananBalachandran) [reply plugins/module_utils/network/sonic/facts/pim_global/pim_global.py:93]: When 'ecmp'/'ecmp_rebalance' is not enabled (default), the data will be empty.
Hence, the enable value for them is set to 'False' if their setting is not available.

File: plugins/module_utils/network/sonic/config/pim_global/pim_global.py [hunk 9]
```diff
@@ -0,0 +1,403 @@
...
+        return requests
+
+    def validate_config(self, want, have):
+        """Validate the given config"""
+        state = self._module.params['state']
+        if state == 'deleted':
+            for conf in want:
+                have_conf = next((cfg for cfg in have if cfg['vrf_name'] == conf['vrf_name']), {})
+                if conf.get('ecmp_enable') and not conf.get('ecmp_rebalance_enable') and have_conf.get('ecmp_rebalance_enable'):
+                    self._module.fail_json(msg='ECMP cannot be disabled when ECMP Rebalance is enabled')
```
[the remark this answers is not in the document]
Comment (ArunSaravananBalachandran) [reply plugins/module_utils/network/sonic/config/pim_global/pim_global.py:365]: As suggested, have updated the check to handle cases for 'replaced' and 'overridden' states.

File: plugins/module_utils/network/sonic/config/pim_global/pim_global.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/pim_global/pim_global.py:287]: Thank you for explaining.

Resolved.

File: plugins/module_utils/network/sonic/facts/pim_global/pim_global.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/facts/pim_global/pim_global.py:93]: I can see why this is needed if we are matching the CLI implementation.

Resolved.

**Comment by kerry-meyer** [comment]
The revised code looks good and all issues are now "Resolved".

The updated test results look good, but there is an anomaly that needs to be investigated in the posted html regression results with 'check and diff' modes. activated:

In 'check' mode, the "result['after']" dictionary should be empty, but it in the html regression results, the 'After' column is being displayed with the same content in the output as the 'After(Generated)' column (except that the dictionary members are in a different order).

The console log shows that, at execution time, the 'after' dict was, as expected, empty. This proves that the correct actions were taken during execution of the resource module for each test case.

Also, the  'After(Generated)' and 'Diff' outputs look correct.

I looked briefly at the corresponding Jinja2 and report generation files to see if there is an explanation there for the unexpected 'After' results in the test reports, but it isn't yet clear to me why this is happening.

Although I think this anomaly requires more investigation, I don't think the problem is in any way related to this change set, so I will proceed with approving and merging the change set.

If you have time, please see if you can determine a reason for the anomaly mentioned here.

**Review by kerry-meyer — APPROVED** [review]
The updated code an corresponding test results all look good.

Thank you very much for implementing the "PIM global" resource module.

Approved.

**Pull request merged**